prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1105 @@
+		(void) util_cancel_entry(srx, flags, rx_entry);
+	}
+	return FI_SUCCESS;
+}
+
+int util_srx_close(struct fid *fid)
+{
+	struct util_srx_ctx *srx;
+	struct util_rx_entry *rx_entry;
+	struct slist_entry *entry;
+
+	srx = container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+	if (!srx)
+		return -FI_EINVAL;
+
+	(void)ofi_array_iter(&srx->src_recv_queues, srx, util_cleanup_queues);
+	(void)ofi_array_iter(&srx->src_trecv_queues, srx, util_cleanup_queues);
+	ofi_array_destroy(&srx->src_recv_queues);
+	ofi_array_destroy(&srx->src_trecv_queues);
+
+	while (!slist_empty(&srx->msg_queue)) {
+		entry = slist_remove_head(&srx->msg_queue);
+		(void) util_cancel_entry(srx, FI_SEND | FI_MSG,
+				container_of(entry, struct util_rx_entry,
+				             peer_entry));
+	}
+
+	while (!slist_empty(&srx->tag_queue)) {
+		entry = slist_remove_head(&srx->tag_queue);
+		(void) util_cancel_entry(srx, FI_SEND | FI_TAGGED,
+				container_of(entry, struct util_rx_entry,
+					     peer_entry));
+	}
+
+	while (!dlist_empty(&srx->all_unexp_msg)) {
+		dlist_pop_front(&srx->all_unexp_msg, struct util_rx_entry,
+				rx_entry, entry);
+		rx_entry->peer_entry.srx->peer_ops->discard_msg(
+							&rx_entry->peer_entry);
+		ofi_buf_free(rx_entry);
+	}
+
+	while (!dlist_empty(&srx->all_unexp_tag)) {
+		dlist_pop_front(&srx->all_unexp_tag, struct util_rx_entry,
+				rx_entry, entry);
+		rx_entry->peer_entry.srx->peer_ops->discard_tag(
+							&rx_entry->peer_entry);
+		ofi_buf_free(rx_entry);
+	}
+
+	ofi_array_destroy(&srx->src_unexp_msg_queues);
+	ofi_array_destroy(&srx->src_unexp_tag_queues);
+
+	ofi_atomic_dec32(&srx->cq->ref);
+	ofi_bufpool_destroy(srx->rx_pool);
+	ofi_mutex_destroy(&srx->lock);
+	free(srx);
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops util_srx_fid_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = util_srx_close,
+	.bind = util_srx_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static bool util_cancel_recv(struct util_srx_ctx *srx, struct slist *queue,
+			     uint64_t flags, void *context)
+{
+	struct slist_entry *item, *prev;
+	struct util_rx_entry *rx_entry;
+
+	slist_foreach(queue, item, prev) {
+		rx_entry = container_of(item, struct util_rx_entry, peer_entry);
+		if (rx_entry->peer_entry.context == context) {
+			slist_remove(queue, item, prev);
+			util_cancel_entry(srx, flags, rx_entry);
+			return true;
+		}
+	}
+	return false;
+}
+
+static int util_cancel_src(struct ofi_dyn_arr *arr, void *list, void *context)
+{
+	struct util_srx_ctx *srx;
+	struct slist *queue = list;
+	uint64_t flags;
+
+	srx = container_of(arr, struct util_srx_ctx, src_trecv_queues);
+
+	flags = arr == &srx->src_trecv_queues ?
+		FI_TAGGED | FI_RECV : FI_MSG | FI_RECV;
+
+	return (int) util_cancel_recv(srx, queue, flags, context);
+}
+
+static ssize_t util_srx_cancel(fid_t ep_fid, void *context)
+{
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	srx->lock_acquire(&srx->lock);
```

Comment:
Why is this the only place where a lock is used?

Reply:
A relevant question: This is being called inside `smr_ep_cancel`, do we still need a ep lock in `smr_ep_cancel` if after we already acquire srx->lock? 

Similar question for generic_recv and ep_progress function. Currently efa acquire both srx lock and ep lock. I know they are different level of locking, so I tend to keep them both. 

@wenduwan  FYI for your comment in efa PR https://github.com/ofiwg/libfabric/pull/8957#discussion_r1210522298